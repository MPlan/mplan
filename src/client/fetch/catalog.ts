import { fetchWithAuth } from './fetch-with-auth';
import * as Model from 'models';

export async function fetchCatalog() {
  const response = await fetchWithAuth('/api/catalog');
  const catalog = (await response.json()) as Model.Catalog.Model;
  return catalog;
}
