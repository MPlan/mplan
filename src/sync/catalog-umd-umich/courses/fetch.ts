import axios from 'axios';

export async function fetchCourseList(level: 'undergraduate' | 'graduate', subjectCode: string) {
  const response = await axios({
    url: `http://catalog.umd.umich.edu/${level}/coursesaz/${subjectCode.toLowerCase().trim()}`,
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });
  const html = response.data as string | undefined;
  if (!html) throw new Error(`could not get course list for "${level} ${subjectCode}"`);
  return html;
}
