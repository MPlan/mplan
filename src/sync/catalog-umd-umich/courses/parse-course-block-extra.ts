export function parseRestrictionsBlock(restrictionsBlock: Element) {
  return Array.from(restrictionsBlock.childNodes)
    .map(x => x.textContent!)
    .filter(x => !!x)
    .map(x => x.trim())
    .filter(x => !!x)
    .filter(x => !/restriction/i.test(x));
}

export function parseCourseBlockExtra(courseBlockExtra: Element) {
  const textContent = courseBlockExtra.textContent;
  if (!textContent) throw new Error('could not get textContent from .courseblockextra');
  const restrictionsBlock = /restriction/i.test(textContent);
  if (restrictionsBlock) return { restrictions: parseRestrictionsBlock(courseBlockExtra) };
  return undefined;
}
