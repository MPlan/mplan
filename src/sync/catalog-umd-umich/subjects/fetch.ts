import axios from 'axios';

export async function fetchUndergraduateSubjectsHtml() {
  const response = await axios({
    method: 'GET',
    url: 'http://catalog.umd.umich.edu/undergraduate/coursesaz/',
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) {
    throw new Error('Subjects response was undefined or empty.');
  }
  return html;
}
