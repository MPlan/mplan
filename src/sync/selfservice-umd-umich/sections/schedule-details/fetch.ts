import axios from 'axios';
import * as https from 'https';
import { oneLineTrim } from 'common-tags';
import { formEncode } from 'utilities/form-encode-decode';

export async function fetchScheduleDetailHtml(termCode: string, crn: string) {
  const url = oneLineTrim`
    https://selfservice.umd.umich.edu/BANP/bwckschd.p_disp_detail_sched?
    ${formEncode({
      term_in: termCode,
      crn_in: crn,
    })}
  `;

  const response = await axios({
    method: 'POST',
    url,
    httpsAgent: new https.Agent({ ciphers: 'ALL' }),
    headers: {
      'X-What-Is-This-Request': 'https://github.com/MPlan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) throw new Error('Subjects response was undefined or empty.');

  return html;
}
