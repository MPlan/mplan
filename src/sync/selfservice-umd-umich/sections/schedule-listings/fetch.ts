import * as https from 'https';
import axios from 'axios';
import { oneLineTrim } from 'common-tags';
import { formEncode } from 'utilities/form-encode-decode';

export async function fetchScheduleListingsHtml(
  termCode: string,
  subjectCode: string,
  courseNumber: string,
) {
  const url = oneLineTrim`
    https://selfservice.umd.umich.edu/BANP/bwckctlg.p_disp_listcrse?
    ${formEncode({
      term_in: termCode.toUpperCase().trim(),
      subj_in: subjectCode.toUpperCase().trim(),
      crse_in: courseNumber.toUpperCase().trim(),
      schd_in: '',
    })}
  `;

  const response = await axios({
    method: 'GET',
    url,
    httpsAgent: new https.Agent({ ciphers: 'ALL' }),
    headers: {
      'X-What-Is-This-Request': 'https://github.com/mplan/mplan',
    },
  });

  const html = response.data as string | undefined;
  if (!html) throw new Error('Subjects response was undefined or empty.');

  return html;
}
