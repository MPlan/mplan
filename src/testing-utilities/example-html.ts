import * as fs from 'fs';

// https://stackoverflow.com/questions/16697791/nodejs-get-filename-of-caller-function
function _getCallerFile(): string {
  const originalFunc = Error.prepareStackTrace;

  let callerFile: any;
  try {
    const err = new Error() as any;
    let currentFile;

    Error.prepareStackTrace = function(err, stack) {
      return stack;
    };

    currentFile = err.stack.shift().getFileName();

    while (err.stack.length) {
      callerFile = err.stack.shift().getFileName();
      if (currentFile !== callerFile) break;
    }
  } catch (e) {}
  Error.prepareStackTrace = originalFunc;
  return callerFile;
}

export function writeExampleHtml(id: string, html: string) {
  const callerFile = _getCallerFile();
  const split = callerFile.split('/');
  const callerDirectory = split.slice(0, split.length - 1).join('/');
  const examplesDirectory = `${callerDirectory}/__examples__`;
  if (!fs.existsSync(examplesDirectory)) fs.mkdirSync(examplesDirectory);
  fs.writeFileSync(`${examplesDirectory}/${id}.html`, html);
}

export function getExampleHtml(id: string) {
  const callerFile = _getCallerFile();
  const split = callerFile.split('/');
  const callerDirectory = split.slice(0, split.length - 1).join('/');
  const examplesDirectory = `${callerDirectory}/__examples__`;
  return fs.readFileSync(`${examplesDirectory}/${id}.html`).toString();
}
