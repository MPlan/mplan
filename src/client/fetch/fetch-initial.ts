import { fetchUser } from './user';
import { fetchCatalog } from './catalog';
import { fetchMasteredDegrees } from './mastered-degrees';

export async function fetchInitial() {
  const [user, catalog, masteredDegrees] = await Promise.all([
    fetchUser(),
    fetchCatalog(),
    fetchMasteredDegrees(),
  ]);
  return { user, catalog, masteredDegrees };
}
