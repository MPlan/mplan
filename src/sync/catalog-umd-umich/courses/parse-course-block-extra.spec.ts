import { JSDOM } from 'jsdom';
import { getOrWriteHtml } from 'testing-utilities/example-html';
import { fetchCourseList } from './fetch';
import { parseRestrictionsBlock } from './parse-course-block-extra';

describe('parseRestrictionsBlock', () => {
  it('parses restrictions and returns an array of them', async () => {
    const html = await getOrWriteHtml('cis-courses', __dirname, () =>
      fetchCourseList('undergraduate', 'cis'),
    );
    const { window } = new JSDOM(html);
    const { document } = window;

    const parsedRestrictions = Array.from(document.querySelectorAll('.courseblockextra'))
      .filter(courseBlockExtra => {
        const textContent = courseBlockExtra.textContent;
        if (!textContent) return false;
        return /restriction/i.test(textContent);
      })
      .map(parseRestrictionsBlock);

    expect(parsedRestrictions).toMatchSnapshot();
  });
});
