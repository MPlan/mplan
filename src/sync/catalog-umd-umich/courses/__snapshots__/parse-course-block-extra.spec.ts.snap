// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseRestrictionsBlock parses restrictions and returns an array of them 1`] = `
Array [
  Array [
    "Can enroll if Program is",
  ],
  Array [
    "Can enroll if Class is Junior or Senior",
  ],
  Array [
    "Cannot enroll if Class is Freshman",
    "Can enroll if Level is Undergraduate",
  ],
  Array [
    "Cannot enroll if Class is Freshman or Sophomore",
  ],
  Array [
    "Can enroll if Class is Junior or Senior",
    "Can enroll if Level is Undergraduate",
  ],
  Array [
    "Can enroll if Class is Junior or Senior",
    "Can enroll if Level is Undergraduate",
    "Can enroll if College is Arts, Sciences, and Letters or Engineering and Computer Science",
  ],
  Array [
    "Cannot enroll if College is Business",
  ],
  Array [
    "Cannot enroll if College is Business or Education, Health, and Human Services",
  ],
  Array [
    "Can enroll if Class is Senior",
    "Can enroll if College is Business",
  ],
  Array [
    "Cannot enroll if Class is Freshman",
    "Cannot enroll if College is Business",
  ],
  Array [
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Cannot enroll if Class is Graduate or Doctorate",
  ],
  Array [
    "Can enroll if Class is Junior or Senior",
    "Can enroll if Level is Undergraduate",
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if Class is Junior or Senior",
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if Class is Senior",
    "Can enroll if Major is CIS/Information Systems",
  ],
  Array [
    "Can enroll if Class is Senior",
    "Can enroll if Major is CIS/Information Systems",
  ],
  Array [
    "Can enroll if Class is Senior",
    "Can enroll if College is Business",
  ],
  Array [
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if College is Engineering and Computer Science",
  ],
  Array [
    "Can enroll if Class is Senior",
  ],
  Array [
    "Can enroll if Class is Senior",
  ],
]
`;
